piptools/_compat/path_compat.py

```diff
@@ -0,0 +1,36 @@
+"""
+Compatibility helpers for working with paths and ``pathlib`` across platforms
+and Python versions.
+"""
+
+from __future__ import annotations
+
+import os.path
+import pathlib
+import sys
+
+
+def relative_to_walk_up(path: pathlib.Path, start: pathlib.Path) -> pathlib.Path:
+    """
+    Compute a relative path allowing for the input to not be a subpath of the start.
+
+    This is a compatibility helper for ``pathlib.Path.relative_to(..., walk_up=True)``
+    on all Python versions. (``walk_up: bool`` is Python 3.12+)
+    """
+    # prefer `pathlib.Path.relative_to` where available
+    if sys.version_info >= (3, 12):
+        return path.relative_to(start, walk_up=True)
+
+    str_result = os.path.relpath(path, start=start)
+    return pathlib.Path(str_result)
+
+
+def is_path_relative_to(path1: pathlib.Path, path2: pathlib.Path) -> bool:
+    """Return True if ``path1`` is relative to ``path2``."""
+    # TODO: remove this function in favor of Path.is_relative_to()
+    #       when we drop support for Python 3.8
```

Comment:
Both 3.8 and 3.9 are both EOL already: https://endoflife.date/python. And testing against pip is becoming difficult in older runtimes.

I think we could plan dropping Python 3.8 once #2257 is implemented and released. With that, there would be no need to even think about hosting this function in compat modules.